index.js

```diff
@@ -274,7 +303,11 @@ class ClinicDoctor extends events.EventEmitter {
       fs.createWriteStream(outputFilename),
       function (err) {
         clearInterval(checkHeapInterval)
-        callback(err)
+        if (err) {
+          callback(err)
+        } else {
+          callback(null, result)
```

Comment:
Having `visualize` output a structured object is a bit odd. If anything, it should output the file stream. A better separation of logic would be `collect() -> analyze() -> visualize()`. Where `visualize()` takes the `analyze()` output. I do see the incompatibility, but maybe `analyze()` should output streams?

I don't have any strong opinions. A valid alternative is also to just force the API consumer to call both methods if they want both a structured object and a file.

Reply:
Yeah, I think `analyze()` → `visualize()` would be ideal, but I didn't wanna break the API :stuck_out_tongue: I did it this way so the CLI could avoid running the analysis twice. I agree with your points tho.